snapshot-service/pb/service.proto

```diff
@@ -0,0 +1,265 @@
+
+option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
+  info: {
+    title: "PLGD Rule Engine API";
+    version: "1.0";
+    description: "API for configuring rule engine in PLGD";
+    contact: {
+      name: "plgd.dev";
+      url: "https://github.com/plgd-dev/hub";
+      email: "info@plgd.dev";
+    };
+    license: {
+      name: "Apache License 2.0";
+      url: "https://github.com/plgd-dev/hub/blob/v2/LICENSE";
+    };
+  };
+  schemes: [HTTPS];
+  consumes: ["application/json", "application/protojson"];
+  produces: ["application/json", "application/protojson"];
+};
+
+option go_package = "github.com/plgd-dev/hub/v2/snapshot-service/pb;pb";
+
+message Id {
+  string id = 1;
+  uint64 version = 2;
+}
+
+message IdFilter {
+  string id = 1;
+  oneof version {
+    uint64 value = 2;
+    bool all = 3;
+    bool max = 4;
+  }
+}
+
+message Condition {
+  Id id = 1;
+  string name = 2;
+  bool enabled = 3;
+  repeated string device_id_filter = 5;
+  repeated string resource_type_filter = 6;
+  repeated string resource_href_filter = 7;
+  string yq_expression = 8;
+  
+  message InvokeConfiguration {
+    string id = 1; // latest version of cfg
+    bool keep_updating_on_failure = 2;
+    string api_access_token = 3; // token used to update resources in the configuration
+    // or do we want to use token from device to update self resources ?
+  }
+  InvokeConfiguration invoke_configuration = 9;
+}
+
+message CreateConditionReponse { Id id = 1; }
+message UpdateConditionResponse {
+  Id id = 1; // for new version 
+} 
+message GetConditionsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConditionsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConditionsResponse { int64 count = 1; }
+
+
+message ResourceTypes {
+  repeated string types = 1; // all types must match resource
+  uint32 min = 2; // minimal number of resources that will be updated
+}
+
+message Configuration {
+  message Resource {
+    oneof filter {
+      string href  = 1;
+      ResourceTypes resource_types = 2; 
+    }
+    resourceaggregate.pb.Content content = 3;
+    string resource_interface = 6; // optional update interface
+    int64 time_to_live = 7; // optional update command time to live
+  }
+  Id id = 1;
+  string name = 2;
+  repeated Resource resources = 3;
+}
+
+message CreateConfigurationReponse { Id id = 1; }
+message UpdateConfigurationReponse  {
+  Id id = 1; // for new version 
+} 
+message GetConfigurationsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConfigurationsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConfigurationsResponse { int64 count = 1; }
+
+message AppliedConfiguration {
+  message Resource {
+    resourceaggregate.pb.ResourceId resource_id = 1;
+    uint32 configuration_resources_idx = 2; // index of resource in configuration resources array. For resource types it could be mutliple resources.
+    enum Status {
```

Comment:
@jkralik what is `WAITING_FOR_RESOURCE` state describing? It would be good to have in system compatible states, mainly if they represent contextually same behaviour. Why not to reuse pending commands states?

Reply:
When resource still not exists in the hub, it is wait for him - it is expected that the resource will be created in short time.

Reply:
That's equal to Pending Command. Isn't it? And timeout is expired for example. Aren't we dealing only with pending commands and their result state?

Reply:
No it's different. Currently you cannot create update resource when the resource changed of the resource is not stored  or it is not published. We can introduce `bool flag` which allows to create update pending request even the resource not exist. If update resource command to aggregate fails we need to somehow report the error.